pkg/spdk/replica.go

```diff
@@ -1543,6 +1609,558 @@ func (r *Replica) SnapshotRangeHashGet(spdkClient *spdkclient.Client, snapshotNa
 	return spdkClient.BdevLvolGetRangeChecksums(spdktypes.GetLvolAlias(r.LvsName, snapLvolName), clusterStartIndex, clusterCount)
 }
 
+// SnapshotCloneDstStart asks the destination replica to start snapshot cloning
+func (r *Replica) SnapshotCloneDstStart(spdkClient *spdkclient.Client, snapshotName, srcReplicaName, srcReplicaAddress string, cloneMode spdkrpc.CloneMode) (err error) {
+	updateRequired := false
+
+	r.Lock()
+	defer func() {
+		r.Unlock()
+
+		if updateRequired {
+			r.UpdateCh <- nil
+		}
+	}()
+
+	if r.isSnapshotCloning {
+		return fmt.Errorf("replica %s cloning is in process", r.Name)
+	}
+	r.isSnapshotCloning = true
+
+	defer func() {
+		if err != nil {
+			if r.State != types.InstanceStateError {
+				r.State = types.InstanceStateError
+			}
+			r.ErrorMsg = err.Error()
+			if r.snapshotCloningDstCache.cloningError == "" {
+				r.snapshotCloningDstCache.cloningError = err.Error()
+				r.snapshotCloningDstCache.cloningState = types.ProgressStateError
+			}
+		} else {
+			if r.State != types.InstanceStateError {
+				r.ErrorMsg = ""
+			}
+		}
+
+		updateRequired = true
+	}()
+
+	// Replica.Delete and Replica.Create do not guarantee that the previous cloning dst replica info is cleaned up
+	if err := r.doCleanupForSnapshotCloneDst(spdkClient, true); err != nil {
+		return errors.Wrapf(err, "failed to clean up the previous cloning dst info for dst replica snapshot "+
+			"clone start, src replica name %s, address %s, snapshot name %s", r.snapshotCloningDstCache.srcReplicaName,
+			r.snapshotCloningDstCache.srcReplicaAddress, r.snapshotCloningDstCache.snapshotName)
+	}
+	// init cloning
+	r.snapshotCloningDstCache.snapshotName = snapshotName
+	r.snapshotCloningDstCache.srcReplicaName = srcReplicaName
+	r.snapshotCloningDstCache.srcReplicaAddress = srcReplicaAddress
+
+	if cloneMode == spdkrpc.CloneMode_CLONE_MODE_SNAPSHOT_LINKED {
+		srcReplicaIP, _, err := splitHostPort(srcReplicaAddress)
+		if err != nil {
+			return errors.Wrapf(err, "failed to split src Replica address %v", srcReplicaAddress)
+		}
+
+		if r.IP != srcReplicaIP {
+			return fmt.Errorf("failed to do snapshot linked-clone: dst replica IP %v is not the same as "+
+				"src replica IP %v", r.IP, srcReplicaIP)
+		}
+
+		srcReplicaServiceCli, err := GetServiceClient(r.snapshotCloningDstCache.srcReplicaAddress)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if errClose := srcReplicaServiceCli.Close(); errClose != nil {
+				r.log.WithError(errClose).Errorf("Failed to close replica %s client with address %s during "+
+					"start cloning at dst", r.snapshotCloningDstCache.srcReplicaName, r.snapshotCloningDstCache.srcReplicaAddress)
+			}
+		}()
+
+		if err := srcReplicaServiceCli.ReplicaSnapshotCloneSrcStart(r.snapshotCloningDstCache.srcReplicaName,
+			snapshotName, r.Name, "", cloneMode); err != nil {
+			return err
+		}
+		r.snapshotCloningDstCache.cloningState = types.ProgressStateComplete
```

Comment:
Don't we need to check the `clone` state before setting `cloningState` to `ProgressStateComplete`?

Reply:
I think no in this case. At the beginning of this function we clean up `r.snapshotCloningDstCache.cloningState`. Then at this point, we know the clone has finished so directly set it to `ProgressStateComplete`. WDYT?